File: node/src/order.ts
import { HttpClient } from './http-client';
import type { ShipzyConfig } from './http-client';
import type { ApiResult } from './index';

// ============ Types ============

export interface OrderListItem {
    id: string;
    order_no: string;
    status: string;
    customer_name?: string;
    total_amount?: number;
    currency?: string;
    created_at: string;
}

export interface OrderListResponse {
    data: OrderListItem[];
    total: number;
    page: number;
    page_size: number;
}

export interface OrderDetail {
    id: string;
    order_no: string;
    status: string;
    customer_name?: string;
    customer_email?: string;
    customer_phone?: string;
    total_amount?: number;
    currency?: string;
    notes?: string;
    created_at: string;
    updated_at: string;
}

// ============ Client ============

export class OrderClient extends HttpClient {
    async list(params: { page?: number; pageSize?: number; status?: string } = {}): Promise<ApiResult<OrderListResponse>> {
        const q = this.buildQuery({ page: params.page, page_size: params.pageSize, status: params.status });
        return this.request(`/api/v1/order/list${q}`);
    }

    async get(id: string): Promise<ApiResult<OrderDetail>> {
        return this.request(`/api/v1/order/${id}`);
    }

    async create(data: Record<string, unknown>): Promise<ApiResult<OrderDetail>> {
        return this.request('/api/v1/order/create', 'POST', data);
    }

    async createWithDocuments(data: Record<string, unknown>): Promise<ApiResult<OrderDetail & { epod_id?: string; ecmr_id?: string }>> {
        return this.request('/api/v1/order/create-with-documents', 'POST', data);
    }

    async update(id: string, data: Record<string, unknown>): Promise<ApiResult<OrderDetail>> {
        return this.request(`/api/v1/order/${id}/update`, 'POST', data);
    }

    async cancel(id: string): Promise<ApiResult<OrderDetail>> {
        return this.request(`/api/v1/order/${id}/cancel`, 'POST', {});
    }
}

File: rust/src/order.rs
use reqwest::Method;

use crate::error::Result;
use crate::http_client::HttpClient;
use crate::types::{OrderListItem, PaginatedResponse};

pub struct OrderClient {
    pub(crate) inner: HttpClient,
}

impl OrderClient {
    pub fn new(inner: HttpClient) -> Self {
        Self { inner }
    }
}

impl OrderClient {
    pub async fn list(
        &self,
        page: i32,
        page_size: i32,
        status: Option<&str>,
    ) -> Result<PaginatedResponse<OrderListItem>> {
        let mut q = format!("?page={}&page_size={}", page, page_size);
        if let Some(s) = status {
            q.push_str(&format!("&status={}", urlencoding::encode(s)));
        }
        self.inner
            .request(&format!("/api/v1/order/list{}", q), Method::GET, None)
            .await
    }

    pub async fn get(&self, id: &str) -> Result<serde_json::Value> {
        self.inner
            .request(
                &format!("/api/v1/order/{}", id),
                Method::GET,
                None,
            )
            .await
    }

    pub async fn create(&self, body: serde_json::Value) -> Result<serde_json::Value> {
        self.inner
            .request("/api/v1/order/create", Method::POST, Some(body))
            .await
    }

    pub async fn create_with_documents(
        &self,
        body: serde_json::Value,
    ) -> Result<serde_json::Value> {
        self.inner
            .request(
                "/api/v1/order/create-with-documents",
                Method::POST,
                Some(body),
            )
            .await
    }

    pub async fn update(
        &self,
        id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value> {
        self.inner
            .request(
                &format!("/api/v1/order/{}/update", id),
                Method::POST,
                Some(body),
            )
            .await
    }

    pub async fn cancel(&self, id: &str) -> Result<serde_json::Value> {
        self.inner
            .request(
                &format!("/api/v1/order/{}/cancel", id),
                Method::POST,
                None,
            )
            .await
    }
}
